cpp/src/arrow/record_batch.h

```diff
@@ -234,6 +234,53 @@ class ARROW_EXPORT RecordBatchReader {
     return batch;
   }
 
+  class RecordBatchReaderIterator {
+   public:
+    RecordBatchReaderIterator() : batch_(RecordBatchEnd()) {}
+
+    explicit RecordBatchReaderIterator(RecordBatchReader* reader)
+        : batch_(RecordBatchEnd()), reader_(reader) {
+      Next();
+    }
+
+    bool operator!=(const RecordBatchReaderIterator& other) const {
+      return batch_ != other.batch_;
+    }
+
+    Result<std::shared_ptr<RecordBatch>> operator*() {
+      ARROW_RETURN_NOT_OK(batch_.status());
+
+      auto batch = std::move(batch_);
+      batch_ = RecordBatchEnd();
+      return batch;
+    }
+
+    RecordBatchReaderIterator& operator++() {
+      Next();
+      return *this;
+    }
+
+   private:
+    std::shared_ptr<RecordBatch> RecordBatchEnd() {
+      return std::shared_ptr<RecordBatch>(NULLPTR);
```

Comment:
This is what `IterationTraits<std::shared_ptr<RecordBatch>>::End()` would have returned, but using it showed the error:
```
incomplete type 'arrow::IterationTraits<std::shared_ptr<arrow::RecordBatch> >' used in nested name specifier
```
which seems due to the abstract nature of the class.